### DISCUSSION TITLE
Gurobi, osqp & qpoases reg

### DISCUSSION THREAD

**Discussion by lakshmip001** [tesseract-robotics/trajopt#271] [General]
Gurobi, OSQP and qpOASES are this installations compulsory for trajopt_ros to work. Because I didnt install any of them but optimization is done when I run the program but crashes. Kindly please help me with the instructions to install Gurobi, osqp & qpoases with ros.

**Comment by mpowelson** [comment]
I can't help with install instructions as I haven't done it yet, but these are not required. BPMPD is included in the repo, so that is what you are seeing being used now.

**Comment by lakshmip001** [comment]
Thanks . I thought my error was because of the optimizer so I thought of try changing it .

**Comment by mpowelson** [comment]
If you're still interested in trying other optimizers, I just ran it with OSQP for the first time. The install instructions on their website worked fine. From [Build From Sources](https://osqp.org/docs/get_started/sources.html#build-from-sources):
```
git clone --recursive https://github.com/oxfordcontrol/osqp
cd osqp
mkdir build
cd build
cmake -G "Unix Makefiles" ..
cmake --build .
sudo cmake --build . --target install
```
Remove via
```
cmake --build . --target uninstall
```

Then you will need to clean and rebuild trajopt
Note: I actually used checkinstall rather than cmake install, but either should work.

**Comment by lakshmip001** [comment]
Thanks.I will try doing it.

**Comment by mpowelson** [comment]
@arocchi Do you have some instructions for using qpOASES? I "installed" using [these instructions](https://projects.coin-or.org/qpOASES/wiki/QpoasesInstallation), and I added these environment variables to my .bashrc file.
```
export TRAJOPT_CONVEX_SOLVER=QPOASES
export qpOASES_SOURCE_DIR=/home/mpowelson/workspaces/temp/ext/qpOASES/qpOASES-3.2.1/src
export qpOASES_BINARY_DIR=/home/mpowelson/workspaces/temp/ext/qpOASES/qpOASES-3.2.1/bin
```
When I rebuild, I get a bunch or errors similar to this
```
/home/mpowelson/workspaces/temp/devel/.private/trajopt_sco/lib/libtrajopt_sco.so: undefined reference to `qpOASES::SQProblem::SQProblem(int, int, qpOASES::HessianType, qpOASES::BooleanType)'
/home/mpowelson/workspaces/temp/devel/.private/trajopt_sco/lib/libtrajopt_sco.so: undefined reference to `qpOASES::QProblem::init(qpOASES::SymmetricMatrix*, double const*, qpOASES::Matrix*, double const*, double const*, double const*, double const*, int&, double*, double const*, double const*, qpOASES::Bounds const*, qpOASES::Constraints const*, double const*)'
/home/mpowelson/workspaces/temp/devel/.private/trajopt_sco/lib/libtrajopt_sco.so: undefined reference to `qpOASES::SparseMatrix::SparseMatrix(int, int, int*, int*, double*)'
/home/mpowelson/workspaces/temp/devel/.private/trajopt_sco/lib/libtrajopt_sco.so: undefined reference to `qpOASES::SparseMatrix::SparseMatrix(int, int, int*, int*, double*)'
/home/mpowelson/workspaces/temp/devel/.private/trajopt_sco/lib/libtrajopt_sco.so: undefined reference to `qpOASES::SQProblem::hotstart(qpOASES::SymmetricMatrix*, double const*, qpOASES::Matrix*, double const*, double const*, double const*, double const*, int&, double*, qpOASES::Bounds const*, qpOASES::Constraints const*)' 
```

**Comment by arocchi** [comment]
@mpowelson it might be a compatibility issue between `3.2.0` and `3.2.1` (it would be weird as it should be a minor release, but looks like a reasonable explanation to me). Could you try with [this repo](https://github.com/arocchi/qpOASES)? This uses `qpOASES` `3.2.0` and is catkinized

**Comment by mpowelson** [comment]
I may have had something else wrong, but the catkinized qpOASES works for me. 

@lakshmip001 Here are instructions to set up qpOASES.

1. Clone [this repo](https://github.com/arocchi/qpOASES/tree/catkinized) into your workspace. Make sure you grab his catkinized branch
2. Add export TRAJOPT_CONVEX_SOLVER=QPOASES to your .bashrc file if you want to set it as the default. Otherwise just set it in the pci
3. clean and rebuild workspace
